File: flutter_client/lib/core/api/api_parts/p2p_channel_util.dart
import '../p2p_rtc_stub.dart' if (dart.library.html) '../p2p_rtc_web.dart';

class P2pChannelUtil {
  static P2pRtcChannel? parseMark(String? raw) {
    final s = (raw ?? '').trim().toLowerCase();
    if (s.isEmpty) return null;
    if (s == 'api') return P2pRtcChannel.api;
    if (s == 'file') return P2pRtcChannel.file;
    if (s == 'upload') return P2pRtcChannel.upload;
    if (s == 'download') return P2pRtcChannel.download;
    if (s == 'video') return P2pRtcChannel.video;
    return null;
  }

  static P2pRtcChannel fallbackForPath(String path) {
    final p = path.trim();
    if (p.startsWith('/api/file/upload')) return P2pRtcChannel.upload;
    if (p.startsWith('/api/file/download')) return P2pRtcChannel.download;
    if (p.startsWith('/api/videoPlayer/')) return P2pRtcChannel.video;
    if (p.startsWith('/api/file/rawFile')) return P2pRtcChannel.file;
    if (p.startsWith('/api/file') || p.startsWith('/api/static')) {
      return P2pRtcChannel.file;
    }
    return P2pRtcChannel.api;
  }

  static String stripP2pChannelFromPath(String path) {
    return stripQueryKey(path, 'p2pChannel');
  }

  static String stripQueryKey(String path, String key) {
    final idx = path.indexOf('?');
    if (idx < 0) return path;
    final base = path.substring(0, idx);
    final query = path.substring(idx + 1);
    if (query.trim().isEmpty) return base;
    final parts = query.split('&');
    final kept = <String>[];
    for (final part in parts) {
      final p = part.trim();
      if (p.isEmpty) continue;
      final eq = p.indexOf('=');
      final k = (eq >= 0 ? p.substring(0, eq) : p).trim();
      if (k == key) continue;
      kept.add(p);
    }
    if (kept.isEmpty) return base;
    return '$base?${kept.join('&')}';
  }

  static ({P2pRtcChannel channel, String path}) resolve({
    required Uri uri,
    P2pRtcChannel? fallbackChannel,
  }) {
    final rawPath = uri.hasQuery ? '${uri.path}?${uri.query}' : uri.path;
    final stripped = stripP2pChannelFromPath(rawPath);
    final mark = parseMark(uri.queryParameters['p2pChannel']);
    final fallback = fallbackChannel ?? fallbackForPath(uri.path);
    return (channel: mark ?? fallback, path: stripped);
  }
}

File: tv_android/app/src/main/java/com/nascabos/tv/core/api/p2p/P2pChannelUtil.kt
package com.nascabos.tv.core.api.p2p

import android.net.Uri

object P2pChannelUtil {
    fun parseMark(raw: String?): P2pRtcChannel? {
        val s = raw?.trim()?.lowercase().orEmpty()
        return when (s) {
            "api" -> P2pRtcChannel.Api
            "file" -> P2pRtcChannel.File
            "upload" -> P2pRtcChannel.Upload
            "download" -> P2pRtcChannel.Download
            "video" -> P2pRtcChannel.Video
            else -> null
        }
    }

    fun fallbackForPath(path: String): P2pRtcChannel {
        val p = path.trim()
        return when {
            p.startsWith("/api/file/upload") -> P2pRtcChannel.Upload
            p.startsWith("/api/file/download") -> P2pRtcChannel.Download
            p.startsWith("/api/videoPlayer/") -> P2pRtcChannel.Video
            p.startsWith("/api/file/rawFile") -> P2pRtcChannel.File
            p.startsWith("/api/file") || p.startsWith("/api/static") -> P2pRtcChannel.File
            else -> P2pRtcChannel.Api
        }
    }

    fun stripP2pChannelFromPath(path: String): String {
        return stripQueryKey(path, "p2pChannel")
    }

    fun stripQueryKey(path: String, key: String): String {
        val idx = path.indexOf('?')
        if (idx < 0) return path
        val base = path.substring(0, idx)
        val query = path.substring(idx + 1)
        if (query.trim().isEmpty()) return base
        val parts = query.split('&')
        val kept = mutableListOf<String>()
        for (part in parts) {
            val p = part.trim()
            if (p.isEmpty()) continue
            val eq = p.indexOf('=')
            val k = (if (eq >= 0) p.substring(0, eq) else p).trim()
            if (k == key) continue
            kept += p
        }
        return if (kept.isEmpty()) base else "$base?${kept.joinToString("&")}"
    }

    data class Resolved(val channel: P2pRtcChannel, val path: String)

    /**
     * 使用 encodedPath/encodedQuery 保留百分号编码，避免文件名含空格、中文等特殊字符时
     * 在 P2P 服务端 new URL(path, base) 报 invalid_url（直连时浏览器/客户端会发已编码 URL，P2P 需原样传递）。
     */
    fun resolve(
        uri: Uri,
        fallbackChannel: P2pRtcChannel? = null,
    ): Resolved {
        val encodedPath = uri.encodedPath?.takeIf { it.isNotEmpty() } ?: ""
        val encodedQuery = uri.encodedQuery
        val rawPath = if (encodedQuery != null && encodedQuery.isNotEmpty()) "$encodedPath?$encodedQuery" else encodedPath
        val stripped = stripP2pChannelFromPath(rawPath)
        val mark = parseMark(uri.getQueryParameter("p2pChannel"))
        val fallback = fallbackChannel ?: fallbackForPath(uri.path ?: "")
        return Resolved(channel = mark ?: fallback, path = stripped)
    }
}
